### PR TITLE
RC-32: GP 'identifierSourceId' to support UUIDs and not only IDs.

### PR DESCRIPTION
**Pull request by samuelmale**
…ncore.identifierSourceId GlobalProperty

Ticket : https://issues.openmrs.org/browse/RC-32

### CODE REVIEW

File: api/src/test/java/org/openmrs/module/registrationcore/api/RegistrationCoreServiceTest.java [hunk 1]
```diff
@@ -282,7 +284,15 @@ public void registerPatient_shouldSucceedIfBiometricsSuppliedAndEngineEnabled()
...
     }
-
+    
+    @Test
+    public void  registerPatient_shouldAlsoConsumeUuidAsValueForIdentifierSourceIdGP() {
+    	String sourceUuid = Context.getService(IdentifierSourceService.class).getIdentifierSource(1).getUuid();
+    	adminService.saveGlobalProperty(new GlobalProperty(RegistrationCoreConstants.GP_OPENMRS_IDENTIFIER_SOURCE_ID, sourceUuid));
+		Location location = locationService.getLocation(2);
+	    service.registerPatient(createBasicPatient(), null, null, location);
+    }
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/registrationcore/api/RegistrationCoreServiceTest.java]: This test doesn't assert anything.

File: api/src/main/resources/moduleApplicationContext.xml [hunk 2]
```diff
@@ -136,4 +136,13 @@
...
           class="org.openmrs.module.registrationcore.api.mpi.openempi.OpenEmpiPatientFilter"/>
+          
+    <!-- Add global property listeners -->
+	<bean id="registrationcore.globalPropertyListeners" parent="openmrsEventListeners">
+	    <property name="globalPropertyListeners">
+	        <list value-type="org.openmrs.api.GlobalPropertyListener" merge="true">
+	            <bean class="${project.parent.groupId}.${project.parent.artifactId}.api.impl.RegistrationCoreServiceImpl" />
+	        </list>
+	    </property>
+	</bean>
```
Comment (mks-d) [inline api/src/main/resources/moduleApplicationContext.xml]: Ok this is coupled with implementing [`RegistrationCoreServiceImpl.globalPropertyChanged(GlobalProperty)`](https://github.com/openmrs/openmrs-module-registrationcore/blob/d9596396ae7d822584dc56c24ea739a35c9a1315/api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java#L249-L252) correct?

If yes, this should be part of another PR I think?

File: api/src/test/java/org/openmrs/module/registrationcore/api/RegistrationCoreServiceTest.java (hunk 1 above)
Comment (samuelmale) [reply api/src/test/java/org/openmrs/module/registrationcore/api/RegistrationCoreServiceTest.java]: Ooohh, I just forgot to do some assertion

File: api/src/main/resources/moduleApplicationContext.xml (hunk 2 above)
Comment (samuelmale) [reply api/src/main/resources/moduleApplicationContext.xml]: We need this even before https://issues.openmrs.org/browse/RC-32 . I trust initial authors just forgot to add it, does it really need creation of a new ticket?

Comment (mks-d) [reply api/src/main/resources/moduleApplicationContext.xml]: Well if you hook the GP listener then you will trigger the below behaviour each time a GP is updated no?
```java
@Override
public void globalPropertyChanged(GlobalProperty gp) {
  idSource = null;
}
```
... and that's unwanted to say the least.

I think a new ticket would be best to handle an appropriate implementation of what should happen when a specific GP's value is changed.

Comment (samuelmale) [reply api/src/main/resources/moduleApplicationContext.xml]: As I think and know, that implementation is appropriate, Even at runtime. I'm sure that implementation is perfect. The only missing thing was adding this config. 

I added this here cause of my test unit. Yes we can leave it out but it's so so crucial. We either create a new ticket and add prior to merging this or let's leave it here @mks-d

Comment (mks-d) [reply api/src/main/resources/moduleApplicationContext.xml]: @samuelmale my understanding is this that, by enabling the GP listener on this class, the above code will be run each time a GP is edited. In other words `idSource` will be set to null as soon as _any_ GP is edited. Are we sure that this is what we want?

I would assume that if you hook the GP listener to that class, then you ought to provide an appropriate implementation for `RegistrationCoreServiceImpl.globalPropertyChanged(GlobalProperty)`. It should check whether the GP that is ediited is `"registrationcore.identifierSourceId"`, and if it is, it should update `idSource` accordingly.

Comment (samuelmale) [reply api/src/main/resources/moduleApplicationContext.xml]: Makes sense. I think that could come after this. Maybe as another ticket! It's actually the right thing. We want to set the `idSource = null` when and only the `registrationcore.identifierSourceId` is edited. This is because the `idSource` is a static member so if the `identifierSource` is changed using the GP the current one will need to be set to null so that the new one is picked up [here](https://github.com/openmrs/openmrs-module-registrationcore/blob/d9596396ae7d822584dc56c24ea739a35c9a1315/api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java#L490-L497). Unless we have the bean above, this won't be possible at runtime once the identifier source was already initialized in the `RegistrationCoreService`.

As for now, I believe the current implementation still works anyway! But anyway, we could create a different ticket for a better implementation of `RegistrationCoreServiceImpl.globalPropertyChanged(GlobalProperty)`  @mks-d

Comment (mks-d) [reply api/src/main/resources/moduleApplicationContext.xml]: @samuelmale ok that's a bit weird, I would rather have a code that ensures that `idSource` is set at the value that it should, rather than having another method safe guarding against the null value.

@mseaton did you get a chance to look at @samuelmale's PR here?

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/IdentifierBuilder.java [hunk 3]
```diff
@@ -37,7 +37,7 @@ public void setPatientService(PatientService patientService) {
      * @param location identifier location
      * @return generated patient identifier
      */
-    public PatientIdentifier generateIdentifier(Integer sourceId, Location location) {
+    public PatientIdentifier generateIdentifier(String sourceId, Location location) {
         location = getLocation(location);
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/registrationcore/api/impl/IdentifierBuilder.java]: I don't know if there are any implications, but this is changing the method signature of a public method.  Might be safer to add a second method that takes a string and performs the necessary logic.

Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/registrationcore/api/impl/IdentifierBuilder.java]: Isn't it safe now, having this more generic. I trust it accepts both String and Integer values @mseaton

Comment (mks-d) [reply api/src/main/java/org/openmrs/module/registrationcore/api/impl/IdentifierBuilder.java]: @samuelmale the problem might be coming from JARs that depend on Idgen and that will complain that this signature is not found anymore:
```java
generateIdentifier(Integer, Location)
```
Those issues can actually be quite nasty to troubleshoot.

As @mseaton said just introduce a new method, keeping the old signature around. In terms of the implementation, don't duplicate code, you will reuse one in the other.

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java [hunk 4]
```diff
@@ -29,9 +29,14 @@ public void setApplicationContext(ApplicationContext applicationContext) {
         this.applicationContext = applicationContext;
     }
 
-    public Integer getOpenMrsIdentifierSourceId() {
+	public Integer getOpenMrsIdentifierSourceId() {
         String propertyName = RegistrationCoreConstants.GP_OPENMRS_IDENTIFIER_SOURCE_ID;
-        return getIntegerProperty(propertyName);
+		return getIntegerProperty(propertyName);
```
Comment (samuelmale) [inline api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java]: I'm sure the above formatting is cause as a result of https://wiki.openmrs.org/display/docs/Developer+How-To+Setup+And+Use+Eclipse .

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java [hunk 5]
```diff
@@ -29,9 +29,14 @@ public void setApplicationContext(ApplicationContext applicationContext) {
...
+	}
+	
+	public String getOpenMrsIdentifierSourceUuid() {
+		String propertyName = RegistrationCoreConstants.GP_OPENMRS_IDENTIFIER_SOURCE_ID;
+        return getProperty(propertyName);
```
Comment (samuelmale) [inline api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java]: Wonder why the formatting turns up so ugly here?

**Comment by samuelmale** [comment]
@mseaton this PR was updated a couple of days back, does it look any better?

**Comment by samuelmale** [comment]
Hi @mseaton, do you have sometime to relook into this, It was updated sometime back as per you objections. Can we have this merged in?

cc: @mks-d

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/IdentifierBuilder.java [hunk 6]
```diff
@@ -37,14 +37,26 @@ public void setPatientService(PatientService patientService) {
...
+	}
+	
+	/**
+	 * Generate new identifier.
+	 * 
+	 * @param sourceId identifier source uuid or id
+	 * @param location identifier location
+	 * @return generated patient identifier
+	 */
+	public PatientIdentifier generateIdentifier(Object sourceId, Location location) {
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/registrationcore/api/impl/IdentifierBuilder.java]: I don't love this.  Having 2 methods with the same name/signature, where one takes in the same arguments as the other is very confusing.  Can you name the method differently so there is no chance of confusion as to which method is hit if you pass in an Integer, Location pair?

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/IdentifierBuilder.java [hunk 7]
```diff
@@ -69,9 +81,15 @@ private Location getLocation(Location identifierLocation) {
         return identifierLocation;
     }
 
-    private IdentifierSource getSource(Integer identifierId) {
-        IdentifierSource identifierSource = getIss().getIdentifierSource(identifierId);
-        validateIdentifierSource(identifierSource, identifierId);
+	private IdentifierSource getSource(Object identifierId) {
+        IdentifierSource identifierSource;
+		String sourceId = identifierId.toString();
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/registrationcore/api/impl/IdentifierBuilder.java]: Null check?

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java (hunk 4 above)
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java]: Did anything change here?

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java (hunk 5 above)
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java]: I find this a little confusing.  If both of these methods exist, I'd prefer to make them both work correctly, regardless of whether an id or uuid is in the GP.  So, in the getOpenMRSIdentifierSourceId method, if it is an integer, return it.  Otherwise, get the source by uuid, and then return the id of the source.  Likewise, in the getOpenMrsIdentifierSourceUuid method, if the property is not an Integer, just return it.  Otherwise, get the source by id, and return the uuid of the source.

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java [hunk 8]
```diff
@@ -488,10 +486,17 @@ private Patient persistImportedMpiPatient(MpiPatient mpiPatient) {
...
+			Object sourceIdentifier = null;
+			try {
+				sourceIdentifier = registrationCoreProperties.getOpenMrsIdentifierSourceId();
+				idSource = iss.getIdentifierSource((Integer) sourceIdentifier);
+			}
+			catch (APIException e) {
+				sourceIdentifier = registrationCoreProperties.getOpenMrsIdentifierSourceUuid();
+				idSource = iss.getIdentifierSourceByUuid((String) sourceIdentifier);
+			}
 			if (idSource == null) {
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java]: If you make the fix I suggest above, you don't need this try/catch here.  Simply get either the uuid or the id, and then return the corresponding source.  Both of those methods should work.

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java (hunk 4 above)
Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java]: Formatting caused as a result of:  https://wiki.openmrs.org/display/docs/Developer+How-To+Setup+And+Use+Eclipse. Planning to add [this](https://github.com/mekomsolutions/openmrs-module-initializer/blob/master/pom.xml#L52-L83) to generally improve on the formatting.

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java (hunk 5 above)
Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java]: Deprecated and kept `getOpenMrsIdentifierSourceId()` for backwards compatibility. We should use `getOpenMrsIdentifierSource()`, hope that clears the confusion

**Comment by samuelmale** [comment]
@mseaton I updated my PR, not sure why the Travis build stops before it completes.

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java (hunk 5 above)
Comment (mseaton) [reply api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java]: Still not sure about this naming.  This doesn't return an identifier source, it returns a String global property value representing either the id or uuid of the source.

Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java]: Ahh, sure.. Nice catch. Let me fix the naming

Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreProperties.java]: @mseaton  corrected :)

**Comment by mseaton** [comment]
@samuelmale looks like you have some conflicts here that need addressing, and we'll need to get the travis build passing before we merge in.

**Comment by samuelmale** [comment]
@mseaton I have resolved the conflicts and I think travis is happy. Thanks for your review.
